README.md
# kodex

Dark nice theme with smooth colors

![Screenshot 0](/screenshots/screenshot0.png)
![Screenshot 1](/screenshots/screenshot1.png)

Arctic Variant

![Screenshot 2](/screenshots/screenshot2.png)
![Screenshot 3](/screenshots/screenshot3.png)

Adopted theme from [kodex Atom](https://github.com/pnx/kodex-syntax) for VS Code.

Based on [OneDark-Pro](https://github.com/Binaryify/OneDark-Pro)

## Author

Henrik Hautakoski - [henrik.hautakoski@gmail.com](mailto:henrik.hautakoski@gmail.com)

CHANGELOG.md
# Change Log

All notable changes to the "kodex" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [0.4.1] - 2023-09-15

### Added

* PHP: Added highlight of `<?php` and `?>` tags.
* PHP: Added highlight of traits.

## [0.4.0] - 2023-06-11

### Changed

- Refactor of all token scopes. trying to first style generic ones like "keyword", "type" and then
  override for specific languages if needed. This results in more consistent theming across languages.

- Variables are now properly highlighted in most cases (as a result of the refactoring).

## [0.3.6] - 2023-06-09

### Changed

- Better highlight for operator tokens.
- Better highlight for some storage type tokens.

## [0.3.5] - 2023-05-28

### Changed

- Better support for "interface" tokens.

## [0.3.4] - 2023-05-25

### Changed

- Better support for "keys" tokens.
- Better support for "function" tokens.

## [0.3.3] - 2023-02-24

### Added

- Support for lua.

## [0.3.2] - 2023-02-20

### Added

- PHP: Added `interface` keyword
- Go: Added `map` keyword
- Added "Operator" syntax class. (default text color)


## [0.3.1] - 2023-02-12

### Changed

- Fix red color for Kodex Arctic

## [0.3.0] - 2023-02-12

### Added

- New variant of the theme: Kodex Arctic!

## [0.2.6] - 2023-02-01

### Added

- Go: syntax highlighting for `chan` keyword
- Go: syntax highlighting for builtin functions.

## [0.2.5] - 2022-11-30

### Added

- Go: syntax highlighting for octal numbers
- Shell: Improved highligting for functions.

## [0.2.4] - 2022-10-30

### Added

- Go: syntax highlighting for `const` keyword.

### Changes

- UI: Panel active border color now uses accent color
- UI: Fix contrast on input options (accent color on background, and light/dark foreground depending on background)

## [0.2.3] - 2022-10-19

### Changed

- Improved tooling.

## [0.2.2] - 2022-10-19

### Added

- TypeScript: Added support for TypeScript again as it got removed by accident in a previous version.

### Changed

- Change yellow color to `#fc9`

## [0.2.1] - 2022-10-19

### Changed

- Ignore development files when packaging extension.

## [0.2.0] - 2022-10-19

### Added

- Template system using typescript.
- Basic python support.

### Changed

- Minor color changes throughout the theme due to the new template system.

## [0.1.5] - 2022-10-15

### Added

- debug icon colors.
- debug console colors.
- TypeScript: Added support for TypeScript

### Changed

- debug status bar colors.
- statusBarItem.hoverBackground to transparent black.

## [0.1.4] - 2022-10-07

### Added

- Golang: highlight hexadecimals as numbers
- JavaScript: support for JavaScript

### Removed

- PHP: highlighting of class name of an inherited class.

## [0.1.3] - 2022-09-29

### Added

- fix `$this` variable and `function` keyword for `php`.

## [0.1.2] - 2022-09-25

### Added

- Highlighting for `golang` and `cmake`.

## [0.1.1] - 2022-09-21

### Added

- Highlighting for `variables` in some languages.
- Support for `php`, `html`, `css` and `scss`.

### Changed

- Overall look-and-feel of the theme is more gray-blueish than just gray.
- Updated terminal colors.

## [0.1.0] - 2022-09-21

- Initial release


src/scheme.ts
import chroma from 'chroma-js'

const colors = {
    // Basic colors
    Black: "#000",
    Gray: "#404754",
    White: "#f1f1f1",
    Blue: "#8ac6f2",
    Cyan: "#8abeb7",
    Green: "#7ccfaf",
    Magenta: "#99c",
    Red: "#cc817f",
    Yellow: "#fc9",
    Orange: "#de935f",

    bg: "#1d232c",
}

const ui = {
    fg: colors.White,
    fgDark: colors.Black,
    bg: colors.bg,
    accent: colors.Blue,
    border: chroma(colors.bg).brighten(0.4).hex(),
    link: chroma(colors.Blue).brighten(0.6).hex(),
    info: {
        fg: colors.Blue,
        bg: chroma(colors.Blue).darken(2).hex(),
        border: chroma(colors.Blue).darken(1.5).hex(),
    },
    success: {
        fg: colors.Green,
        bg: chroma(colors.Green).darken(1).hex(),
        border: colors.Green
    },
    warning: {
        fg: colors.Yellow,
        bg: colors.Orange,
        border: chroma(colors.Orange).darken(1).hex()
    },
    error: {
        fg: colors.Red,
        bg: chroma(colors.Red).darken(2).hex(),
        border: chroma(colors.Red).darken(1.5).hex()
    }
}

const vcs = {
    added: colors.Green,
    deleted: colors.Red,
    modified: colors.Yellow,
    untracked: colors.Blue,
    ignored: chroma(colors.White).darken(2).hex()
}

const diff = {
    added: vcs.added,
    deleted: vcs.deleted,
    modified: vcs.modified,
    fill: colors.Blue,
}

const syntax = {
    tag: colors.Magenta,
    function: colors.Yellow,
    type: colors.Blue,
    string: colors.Magenta,
    markup: "",
    keyword: colors.Yellow,
    variable: colors.Green,
    comment: colors.Red,
    constant: colors.Red,
    number: colors.Red,
    operator: colors.White,
}

const terminal = {
    black: colors.Black,
    white: colors.White,
    blue: colors.Blue,
    cyan: colors.Cyan,
    green: colors.Green,
    magenta: colors.Magenta,
    red: colors.Red,
    yellow: colors.Yellow
}

export default {
    ui,
    vcs,
    diff,
    syntax,
    terminal
}

src/theme_arctic.ts
import chroma from 'chroma-js'

const colors = {
    // Basic colors
    Black: "#000",
    Gray: "#404754",
    White: "#f1f1f1",
    Blue: "#8ac6f2",
    Cyan: "#8abeb7",
    Green: "#7ccfaf",
    Magenta: "#99c",
    Red: "#cc817f",
    Yellow: "#fc9",
    Orange: "#de935f",

    bg: "#282a33",
}

const ui = {
    fg: colors.White,
    fgDark: colors.Black,
    bg: colors.bg,
    accent: colors.Blue,
    border: chroma(colors.bg).brighten(0.4).hex(),
    link: chroma(colors.Blue).brighten(0.6).hex(),
    info: {
        fg: colors.Blue,
        bg: chroma(colors.Blue).darken(2).hex(),
        border: chroma(colors.Blue).darken(1.5).hex(),
    },
    success: {
        fg: colors.Green,
        bg: chroma(colors.Green).darken(1).hex(),
        border: colors.Green
    },
    warning: {
        fg: colors.Yellow,
        bg: colors.Orange,
        border: chroma(colors.Orange).darken(1).hex()
    },
    error: {
        fg: colors.Red,
        bg: chroma(colors.Red).darken(2).hex(),
        border: chroma(colors.Red).darken(1.5).hex()
    }
}

const vcs = {
    added: colors.Green,
    deleted: colors.Red,
    modified: colors.Yellow,
    untracked: colors.Blue,
    ignored: chroma(colors.White).darken(2).hex()
}

const diff = {
    added: vcs.added,
    deleted: vcs.deleted,
    modified: vcs.modified,
    fill: colors.Blue,
}

const syntax = {
    tag: colors.Magenta,
    function: colors.Yellow,
    type: colors.Blue,
    string: colors.Magenta,
    markup: "",
    keyword: colors.Yellow,
    variable: colors.Green,
    comment: colors.Red,
    constant: colors.Red,
    number: colors.Red,
    operator: colors.White,
}

const terminal = {
    black: colors.Black,
    white: colors.White,
    blue: colors.Blue,
    cyan: colors.Cyan,
    green: colors.Green,
    magenta: colors.Magenta,
    red: colors.Red,
    yellow: colors.Yellow
}

export default {
    ui,
    vcs,
    diff,
    syntax,
    terminal
}

src/template.ts
import chroma from 'chroma-js'
import * as helper from './color_helpers'

// https://code.visualstudio.com/api/references/theme-color

export default (name: string, scheme: any) => {

    // Helper variables.
    const fgDarken = chroma(scheme.ui.fg).darken().hex()

    const bg = [
        chroma(scheme.ui.bg).darken(0.25).hex(),
        chroma(scheme.ui.bg).darken(0.2).hex(),
        chroma(scheme.ui.bg).darken(0.1).hex(),
        scheme.ui.bg,
        chroma(scheme.ui.bg).brighten(0.2).hex(),
        chroma(scheme.ui.bg).brighten(0.4).hex(),
        chroma(scheme.ui.bg).brighten(1).hex(),
        chroma(scheme.ui.bg).brighten(1.4).hex()
    ]

    return {
        name: name,
        colors: {

            // Base
            "foreground": scheme.ui.fg,
            "focusBorder": scheme.ui.border,

            // Notifications
            "notificationCenter.border": scheme.ui.border,
            "notificationCenterHeader.foreground": scheme.ui.fg,
            "notificationCenterHeader.background": scheme.ui.accent,
            "notificationToast.border": scheme.ui.border,
            "notifications.foreground": scheme.ui.fg,
            "notifications.background": bg[4],

            // Editor
            "editor.background": bg[3],
            "editor.foreground": scheme.ui.fg,
            "editor.lineHighlightBackground": bg[4],
            "editor.selectionBackground": bg[5],
            "editor.selectionHighlightBorder": scheme.ui.fg,

            // Editor bracket
            "editorBracketMatch.background": bg[7],
            "editorBracketMatch.border": scheme.ui.border,
            "editorBracketHighlight.foreground1": scheme.ui.fg,
            "editorBracketHighlight.foreground2": scheme.ui.fg,
            "editorBracketHighlight.foreground3": scheme.ui.fg,

            // Editor cursor
            "editorCursor.foreground": scheme.ui.fg,

            "editorGroupHeader.tabsBackground": bg[1],
            "editorHoverWidget.background": bg[4],
            "editorHoverWidget.foreground": helper.ContrastColor(bg[4], scheme.ui.fg, scheme.ui.fgDark),
            "editorHoverWidget.border": scheme.ui.border,
            "editorSuggestWidget.background": bg[4],
            "editorSuggestWidget.border": scheme.ui.border,
            "editorSuggestWidget.selectedBackground": bg[5],
            "editorWarning.foreground": scheme.ui.warning.fg,
            "editorWhitespace.foreground": chroma(scheme.ui.fg).alpha(0.1).hex(),

            // Editor indent
            "editorIndentGuide.activeBackground": chroma(scheme.ui.border).brighten(1).hex(),
            "editorIndentGuide.background": scheme.ui.border,

            // Editor find
            "editor.findMatchBackground": scheme.ui.info.bg,
            "editor.findMatchBorder": scheme.ui.info.border,
            "editor.findMatchHighlightBackground": scheme.ui.info.bg,

            // Editor widget
            "editorWidget.background": bg[1],

            // Activity bar
            "activityBar.background": bg[0],
            "activityBar.foreground": scheme.ui.accent,
            "activityBar.activeBackground": chroma(scheme.ui.accent).alpha(0.25).hex(),
            "activityBar.inactiveForeground": scheme.ui.fg,
            "activityBarBadge.background": scheme.ui.accent,
            "activityBarBadge.foreground": helper.ContrastColor(scheme.ui.accent, scheme.ui.fg, scheme.ui.fgDark),

            // Text
            "textLink.foreground": scheme.ui.link,
            "textBlockQuote.background": bg[0],
            "textBlockQuote.border": scheme.ui.border,
            "textCodeBlock.background": bg[0],
            "textPreformat.foreground": scheme.ui.accent,

            // Diff
            "diffEditor.insertedTextBackground": chroma(scheme.diff.added).alpha(0.4).hex(),
            "diffEditor.insertedLineBackground": chroma(scheme.diff.added).alpha(0.3).hex(),
            "diffEditor.removedLineBackground": chroma(scheme.diff.deleted).alpha(0.3).hex(),
            "diffEditor.diagonalFill": chroma(scheme.diff.fill).alpha(0.2).hex(),

            // Line number
            "editorLineNumber.activeForeground": scheme.ui.accent,
            "editorLineNumber.foreground": chroma(scheme.ui.accent).luminance(0.15).desaturate(0.1).hex(),

            // Gutter
            "editorGutter.background": bg[2],
            "editorGutter.addedBackground": scheme.diff.added,
            "editorGutter.deletedBackground": scheme.diff.deleted,
            "editorGutter.modifiedBackground": scheme.diff.modified,

            // Debug
            "editor.stackFrameHighlightBackground": chroma(scheme.ui.warning.bg).alpha(.4).hex(),
            "debugIcon.startForeground": scheme.ui.success.fg,
            "debugIcon.stopForeground": scheme.ui.error.fg,
            "debugIcon.restartForeground": scheme.ui.warning.fg,
            "debugIcon.pauseForeground": scheme.ui.info.fg,
            "debugIcon.continueForeground": scheme.ui.success.fg,
            "debugIcon.stepBackForeground": scheme.ui.info.fg,
            "debugIcon.stepIntoForeground": scheme.ui.info.fg,
            "debugIcon.stepOutForeground": scheme.ui.info.fg,
            "debugIcon.breakpointCurrentStackframeForeground": scheme.ui.warning.fg,
            "debugIcon.breakpointForeground": scheme.ui.error.fg,

            // Debug Toolbar
            "debugToolBar.background": bg[1],

            // Debug Console
            "debugConsole.infoForeground": scheme.ui.info.fg,
            "debugConsole.errorForeground": scheme.ui.error.fg,
            "debugConsole.warningForeground": scheme.ui.warning.fg,

            // Debug View
            "debugView.valueChangedHighlight": scheme.ui.warning.bg,

            // Debug Tokens
            "debugTokenExpression.name": scheme.syntax.tag,
            "debugTokenExpression.value": fgDarken,
            "debugTokenExpression.string": scheme.syntax.string,
            "debugTokenExpression.boolean": scheme.syntax.constant,
            "debugTokenExpression.number": scheme.syntax.number,

            // Testing
            "testing.iconFailed": scheme.ui.error.fg,
            "testing.iconErrored": scheme.ui.error.fg,
            "testing.iconPassed": scheme.ui.success.fg,
            "testing.message.info.lineBackground": scheme.ui.info.bg,
            "testing.message.error.lineBackground": scheme.ui.error.bg,

            // Menu
            "menu.background": bg[1],
            "menu.foreground": scheme.ui.fg,
            "menu.selectionBackground": scheme.ui.bg,
            "menu.separatorBackground": scheme.ui.border,
            "menubar.selectionBackground": scheme.ui.bg,
            "menubar.selectionForeground": scheme.ui.fg,

            // Input
            "input.background": bg[4],
            "input.foreground": scheme.ui.fg,
            "input.border": scheme.ui.border,
            "inputOption.hoverBackground": bg[6],
            "inputOption.activeForeground": helper.ContrastColor(scheme.ui.accent, scheme.ui.fg, scheme.ui.fgDark),
            "inputOption.activeBackground": scheme.ui.accent,
            "inputOption.activeBorder": "#00000000",
            "inputValidation.infoBackground": scheme.ui.info.bg,
            "inputValidation.infoForeground": scheme.ui.info.fg,
            "inputValidation.infoBorder": scheme.ui.info.border,
            "inputValidation.warningBackground": scheme.ui.warning.bg,
            "inputValidation.warningForeground": scheme.ui.warning.fg,
            "inputValidation.warningBorder": scheme.ui.warning.border,
            "inputValidation.errorBackground": scheme.ui.error.bg,
            "inputValidation.errorForeground": scheme.ui.error.fg,
            "inputValidation.errorBorder": scheme.ui.error.border,

            // Button
            "button.background": scheme.ui.accent,
            "button.foreground": helper.ContrastColor(scheme.ui.accent, scheme.ui.fg, scheme.ui.fgDark),
            "button.secondaryBackground": bg[5],
            "button.secondaryForeground": helper.ContrastColor(bg[5], scheme.ui.fg, scheme.ui.fgDark),

            // Dropdown
            "dropdown.background": bg[4],
            "dropdown.border": scheme.ui.border,
            "dropdown.listBackground": scheme.ui.bg,

            // Checkbox
            "checkbox.foreground": scheme.ui.accent,
            "checkbox.background": bg[4],
            "checkbox.border": scheme.ui.border,

            // Badge
            "badge.background": bg[5],
            "badge.foreground": helper.ContrastColor(bg[5], scheme.ui.fg, scheme.ui.fgDark),

            // Lists
            "list.activeSelectionBackground": bg[5],
            "list.activeSelectionForeground": scheme.ui.fg,
            "list.focusBackground": bg[4],
            "list.focusForeground": scheme.ui.fg,
            "list.highlightForeground": scheme.ui.fg,
            "list.hoverBackground": bg[4],
            "list.hoverForeground": scheme.ui.fg,
            "list.inactiveSelectionBackground": bg[4],
            "list.inactiveSelectionForeground": scheme.ui.fg,
            "list.warningForeground": scheme.ui.warning.fg,

            // Minimap gutter
            "minimapGutter.addedBackground": scheme.diff.added,
            "minimapGutter.deletedBackground": scheme.diff.deleted,
            "minimapGutter.modifiedBackground": scheme.diff.modified,

            // Panel
            "panel.border": scheme.ui.border,
            "panel.background": bg[3],
            "panelTitle.activeBorder": scheme.ui.accent,

            // Peak View
            "peekView.border": scheme.ui.accent,
            "peekViewTitle.background": chroma(scheme.ui.accent).alpha(.2).hex(),
            "peekViewTitleDescription.foreground": fgDarken,
            "peekViewTitleLabel.foreground": scheme.ui.fg,
            "peekViewEditor.background": bg[4],
            "peekViewEditorGutter.background": bg[3],
            "peekViewEditor.matchHighlightBackground": bg[6],
            "peekViewResult.background": bg[4],

            // Scrollbar
            "scrollbar.shadow": chroma(scheme.ui.bg).brighten(2).hex(),
            "scrollbarSlider.activeBackground": chroma(scheme.ui.bg).brighten().alpha(0.7).hex(),
            "scrollbarSlider.background": chroma(scheme.ui.bg).brighten().alpha(0.7).hex(),
            "scrollbarSlider.hoverBackground": chroma(scheme.ui.bg).brighten().alpha(0.7).hex(),

            // Sidebar
            "sideBar.background": bg[1],
            "sideBar.foreground": scheme.ui.fg,
            "sideBarSectionHeader.background": chroma(scheme.ui.bg).darken(0.3).hex(),
            "sideBarSectionHeader.foreground": scheme.ui.fg,

            // Statusbar
            "statusBar.background": bg[1],
            "statusBar.foreground": helper.ContrastColor(bg[1], chroma(scheme.ui.fg).darken().hex(), scheme.ui.fgDark),
            "statusBar.noFolderBackground": "#21252b",
            "statusBar.focusBorder": "#212934",
            "statusBar.debuggingBackground": scheme.ui.warning.bg,
            "statusBar.debuggingForeground": helper.ContrastColor(scheme.ui.warning.bg, scheme.ui.fg, scheme.ui.fgDark),
            "statusBar.debuggingBorder": "#00000000",
            "statusBarItem.hoverBackground": chroma("#000000").alpha(0.2).hex(),
            "statusBarItem.remoteBackground": scheme.ui.accent,
            "statusBarItem.remoteForeground": "#000",
            "statusBarItem.prominentBackground": "#8ac6f2",
            "statusBarItem.prominentForeground": "#000",

            // Tabs
            "tab.activeBackground": bg[4],
            "tab.activeBorder": scheme.ui.accent,
            "tab.activeForeground": scheme.ui.fg,
            "tab.border": scheme.ui.bg,
            "tab.hoverBackground":  chroma(scheme.ui.bg).brighten(0.15).hex(),
            "tab.inactiveBackground": scheme.ui.bg,
            // "tab.unfocusedHoverBackground": "#323842",

            // Terminal
            "terminal.ansiBlack": scheme.terminal.black,
            "terminal.ansiWhite": scheme.terminal.white,
            "terminal.ansiBlue": scheme.terminal.blue,
            "terminal.ansiCyan": scheme.terminal.cyan,
            "terminal.ansiGreen": scheme.terminal.green,
            "terminal.ansiMagenta": scheme.terminal.magenta,
            "terminal.ansiRed": scheme.terminal.red,
            "terminal.ansiYellow": scheme.terminal.yellow,
            "terminal.ansiBrightBlack": scheme.terminal.black,
            "terminal.ansiBrightWhite": scheme.terminal.white,
            "terminal.ansiBrightBlue": scheme.terminal.blue,
            "terminal.ansiBrightCyan": scheme.terminal.cyan,
            "terminal.ansiBrightGreen": scheme.terminal.green,
            "terminal.ansiBrightMagenta": scheme.terminal.magenta,
            "terminal.ansiBrightRed": scheme.terminal.red,
            "terminal.ansiBrightYellow": scheme.terminal.yellow,
            "terminal.background": scheme.ui.bg,
            "terminal.border": scheme.ui.border,
            "terminal.foreground": fgDarken,
            "terminal.selectionBackground": chroma(scheme.ui.fg).alpha(0.25).hex(),

            // Titlebar
            "titleBar.activeBackground": bg[1],
            "titleBar.activeForeground": scheme.ui.fg,
            "titleBar.inactiveBackground": bg[1],
            "titleBar.inactiveForeground": fgDarken,

            // Settings
            "settings.focusedRowBackground": bg[5],
            "settings.modifiedItemIndicator": scheme.ui.accent,

            // Git
            "gitDecoration.ignoredResourceForeground": scheme.vcs.ignored,
            "gitDecoration.addedResourceForeground": scheme.vcs.added,
            "gitDecoration.modifiedResourceForeground": scheme.vcs.modified,
            "gitDecoration.deletedResourceForeground": scheme.vcs.deleted,
            "gitDecoration.untrackedResourceForeground": scheme.vcs.untracked,

            // Tree
            "tree.indentGuidesStroke": chroma(scheme.ui.fg).alpha(0.25).hex(),
        },
        tokenColors: [
            {
                "name": "String",
                "scope": [
                    "string",
                    "variable.parameter.url"
                ],
                "settings": {
                    "foreground": scheme.syntax.string
                }
            },
            {
                "name": "Tag",
                "scope": [
                    "constant.other.symbol",
                    "constant.other.key",
                    "markup.heading",
                    "markup.inserted.git_gutter",
                    "meta.group.braces.curly constant.other.object.key.js string.unquoted.label.js",
                    "punctuation.section.embedded.begin.php",
                    "punctuation.section.embedded.end.php"
                ],
                "settings": {
                    "foreground": scheme.syntax.tag
                }
            },
            {
                "name": "Numbers",
                "scope": [
                    "constant.numeric",
                    "keyword.other.unit.hexadecimal.go",
                    "keyword.other.unit.octal.go"
                ],
                "settings": {
                    "foreground": scheme.syntax.number
                }
            },
            {
                "name": "Constants",
                "scope": [
                    "constant.character",
                    "constant.source.cmake"
                ],
                "settings": {
                    "foreground": scheme.syntax.constant
                }
            },
            {
                "name": "Types, Macros, Character Escape, Special Tags",
                "scope": [
                    "storage.type",
                    "meta.type",
                    "meta.return.type",
                    "entity.name.type",

                    // Language specific
                    "keyword.interface.go",
                    "keyword.channel.go",
                    "keyword.local.lua",
                    "keyword.blade",
                ],
                "settings": {
                    "foreground": scheme.syntax.type
                }
            },
            {
                "name": "Comment",
                "scope": [
                    "comment",
                    "punctuation.definition.comment"
                ],
                "settings": {
                    "foreground": scheme.syntax.comment
                }
            },
            {
                "name": "Keyword, Namespace",
                "scope": [
                    "keyword",
                    "keyword.operator.type",
                    "keyword.operator.new",
                    "keyword.operator.expression",

                    "entity.name.namespace",
                    "entity.name.tag",

                    "punctuation.definition.evaluation.parens",

                    "storage.modifier",
                    "storage.type.class",
                    "storage.type.function",
                    "storage.type.interface",
                    "storage.type.type",
                    "storage.type.namespace",
                    "support.type.property-name",

                    // Language specific
                    "storage.type.ts",
                    "storage.type.js",
                    "storage.type.rust",
                    "storage.type.trait.php",
                ],
                "settings": {
                    "foreground": scheme.syntax.keyword
                }
            },
            {
                "name": "Operator",
                "scope": [
                    "keyword.operator"
                ],
                "settings": {
                    "foreground": scheme.syntax.operator
                }
            },
            {
                "name": "Variables",
                "scope": [
                    "variable",
                    "punctuation.definition.variable",
                ],
                "settings": {
                    "foreground": scheme.syntax.variable
                }
            },
            {
                "name": "Invalid",
                "scope": [
                    "invalid",
                    "invalid.illegal"
                ],
                "settings": {
                    "foreground": "#FF0000",
                    "fontStyle": "underline"
                }
            },
            {
                "name:": "No color",
                "scope": [
                    "entity.name.type.class",
                    "entity.name.type.interface",
                    "entity.name.type.module",
                    "entity.name.type.trait.php",
                ],
                "settings": {
                    "foreground": scheme.ui.fg,
                    "fontStyle": "normal"
                },
            }
        ]
    }
}